### PR TITLE
feat(ord): Support equality of StructArray

### LINKED ISSUE #5199: Support equality of StructArray
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

In [datafusion #6635](https://github.com/apache/arrow-datafusion/issues/6635), we think convert tuple to struct will be helpful to implement multi column In list. 

But StructAray is not comparable in arrow-rs since it is nested.

https://github.com/apache/arrow-rs/blob/7e289134a8d9f46a92a2759a7b2488b17993fd5b/arrow-ord/src/cmp.rs#L201-L205.

**Describe the solution you'd like**

I don't know if this feature is already implemented in arrow.

In my view,  compare PrimitiveArray may be rather different from NestedArray. Take equality of StrutArray as an example, I think it may need those steps:

- check equality of datatype
- check equality of PrimitiveArray for each field
- perform conjunction operation on the results of each field

The process slightly differs from the ```apply_op``` process.

https://github.com/apache/arrow-rs/blob/7e289134a8d9f46a92a2759a7b2488b17993fd5b/arrow-ord/src/cmp.rs#L361-L364

**Describe alternatives you've considered**

**Additional context**

The equality of other array type may need more further consideration. Like equality for ListArray, shall we considerate out of order between list(which means that whether [[1,2]] and [[2,1]] are equal or not)?

### PR DESCRIPTION
**Pull request by my-vegetable-has-exploded**
# Which issue does this PR close?

Closes #5199 

# Rationale for this change
 

# What changes are included in this PR?

- refactor closure ```values()``` to function ```compare_op_struct_values()```
- compare struct arrays by recursively checking each field

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-ord/src/cmp.rs [hunk 1]
```diff
@@ -169,12 +169,14 @@ pub fn not_distinct(lhs: &dyn Datum, rhs: &dyn Datum) -> Result<BooleanArray, Ar
 /// Perform `op` on the provided `Datum`
 #[inline(never)]
 fn compare_op(op: Op, lhs: &dyn Datum, rhs: &dyn Datum) -> Result<BooleanArray, ArrowError> {
-    use arrow_schema::DataType::*;
-    let (l, l_s) = lhs.get();
-    let (r, r_s) = rhs.get();
+    let (l_array, l_s) = lhs.get();
```
Comment (my-vegetable-has-exploded) [inline arrow-ord/src/cmp.rs:172]: I renamed (l、r) to (l_array、r_array), so them won't conflict with (l、r) on line 226

**Comment by my-vegetable-has-exploded** [comment]
And there are some points that I am not sure about

- for stuctarray,  a struct array has its own validity bitmap that is independent of its child arrays’ validity bitmaps.  So I don't handle nullbuffer for each field.
- I don't find a way to new a structscalar, so I don't test scalar yet.

Maybe another question, how can I make these test codes shorter?

File: arrow-ord/src/cmp.rs [hunk 2]
```diff
@@ -268,23 +240,122 @@ fn compare_op(op: Op, lhs: &dyn Datum, rhs: &dyn Datum) -> Result<BooleanArray,
...
+) -> Result<BooleanBuffer, ArrowError> {
+    // when one of field is equal, the result is false for not equal
+    // so we use neg to reverse the result of equal when handle not equal
+    let neg = match op {
+        Op::Equal => false,
+        Op::NotEqual => true,
+        _ => unreachable!(),
+    };
+
+    let l = l.as_any().downcast_ref::<StructArray>().unwrap();
```
Comment (jayzhan211) [inline arrow-ord/src/cmp.rs:338]: nit: probably `l.as_struct` can be used here

Comment (my-vegetable-has-exploded) [reply arrow-ord/src/cmp.rs]: get it, thanks @jayzhan211

File: arrow-ord/src/cmp.rs (hunk 1 above)
Comment (alamb) [reply arrow-ord/src/cmp.rs:172]: Thank you -- I find the new naming easier to follow

File: arrow-ord/src/cmp.rs [hunk 3]
```diff
@@ -268,23 +240,119 @@ fn compare_op(op: Op, lhs: &dyn Datum, rhs: &dyn Datum) -> Result<BooleanArray,
...
+    op: Op,
+    l: &dyn Array,
+    l_s: bool,
+    r: &dyn Array,
+    r_s: bool,
+    len: usize,
+) -> Result<BooleanBuffer, ArrowError> {
+    // when one of field is equal, the result is false for not equal
+    // so we use neg to reverse the result of equal when handle not equal
+    let neg = match op {
```
Comment (alamb) [inline arrow-ord/src/cmp.rs:332]: Shouldn't this code also handle `Op::Distinct` and `Op::NotDistinct` (it seems like they would call through to this code as well)

File: arrow-ord/src/cmp.rs [hunk 4]
```diff
@@ -702,4 +772,122 @@ mod tests {
...
+        assert_eq!(eq(&left_struct, &right_struct).unwrap(), expected);
+        assert_eq!(eq(&right_struct, &left_struct).unwrap(), expected);
+        let expected = BooleanArray::new(
+            vec![false, true, false, false].into(),
+            Some(vec![true, true, true, false].into()),
+        );
+        assert_eq!(neq(&left_struct, &right_struct).unwrap(), expected);
+        assert_eq!(neq(&right_struct, &left_struct).unwrap(), expected);
+
+        // test struct('a' , struct('suba', 'subb')) = struct('a', struct('suba', 'subb')), where the right suba1[1] different from left suba[1],the null buffer is 0b0111
```
Comment (alamb) [inline arrow-ord/src/cmp.rs:851]: 👌  very nice

File: arrow-ord/src/cmp.rs [hunk 5]
```diff
@@ -268,23 +240,119 @@ fn compare_op(op: Op, lhs: &dyn Datum, rhs: &dyn Datum) -> Result<BooleanArray,
...
+    let r = r.as_struct();
+
+    // compare each field of struct
+    let child_res = l
+        .columns()
+        .to_vec()
+        .iter()
+        .zip(r.columns().to_vec().iter())
+        .map(|(col_l, col_r)| compare_op_values(Op::Equal, col_l, l_s, col_r, r_s, len))
+        .collect::<Result<Vec<BooleanBuffer>, ArrowError>>()?;
```
Comment (alamb) [inline arrow-ord/src/cmp.rs:348]: I don't think you need to make a new `Vec` here. This compiled for me:

```suggestion
    let child_res = l
        .columns()
        .iter()
        .zip(r.columns().iter())
        .map(|(col_l, col_r)| compare_op_values(Op::Equal, col_l, l_s, col_r, r_s, len))
        .collect::<Result<Vec<BooleanBuffer>, ArrowError>>()?;
```

File: arrow-ord/src/cmp.rs [hunk 6]
```diff
@@ -702,4 +772,122 @@ mod tests {
...
+
+    #[test]
+    fn test_struct_equality() {
+        // test struct('a', 'b') = struct('a', 'b'), the null buffer is 0b0111
+        let left_a = Arc::new(Int32Array::new(
+            vec![0, 1, 2, 3].into(),
+            Some(vec![true, false, true, false].into()),
+        ));
+        let right_a = Arc::new(Int32Array::new(
+            vec![0, 1, 2, 3].into(),
```
Comment (alamb) [inline arrow-ord/src/cmp.rs:784]: I think the test coverage would be improved if `right_a` and `left_a` had some different values 

Like maybe

```suggestion
            vec![0, 1, 20, 3].into(),
```

And a similar change below for `right_b` `left_b`

**Review by alamb — COMMENTED** [review]
Thank you @my-vegetable-has-exploded  -- this is looking pretty close to me

Can you please add tests for
1. `distinct` / `not_distinct`
2. A negative test that some operation like `lt` or `lt_eq` returns an error (not a panic) for struct arrays?
3. A negative test that a struct array like `{a: int, b:int}` doesn't return `true` when compared to a struct array with a prefix like `{a:int}


Also @tustvold do you have any suggestions for what benchmarks to run this on?

**Comment by tustvold** [comment]
> So I don't handle nullbuffer for each field.

I think the output should be the union of all the null buffers.

I'll try to review this in the next couple of days

**Comment by my-vegetable-has-exploded** [comment]
> I think the output should be the union of all the null buffers.

I think the nullbuffer for subarrays is only valid for the subarray itself.

Take the Example Layout in the documentation as an example(https://arrow.apache.org/docs/format/Columnar.html#struct-layout),  if use the union of all the null buffers,  the second slot also gets null, which is a little different from my understanding.

```
[{'joe', 1}, {null, 2}, null, {'mark', 4}]

* Length: 4, Null count: 1
* Validity bitmap buffer:

  | Byte 0 (validity bitmap) | Bytes 1-63            |
  |--------------------------|-----------------------|
  | 00001011                 | 0 (padding)           |

* Children arrays:
  * field-0 array (`VarBinary`):
    * Length: 4, Null count: 2
    * Validity bitmap buffer:

      | Byte 0 (validity bitmap) | Bytes 1-63            |
      |--------------------------|-----------------------|
      | 00001001                 | 0 (padding)           |

    * Offsets buffer:

      | Bytes 0-19     | Bytes 20-63           |
      |----------------|-----------------------|
      | 0, 3, 3, 3, 7  | unspecified (padding) |

     * Value buffer:

      | Bytes 0-6      | Bytes 7-63            |
      |----------------|-----------------------|
      | joemark        | unspecified (padding) |

  * field-1 array (int32 array):
    * Length: 4, Null count: 1
    * Validity bitmap buffer:

      | Byte 0 (validity bitmap) | Bytes 1-63            |
      |--------------------------|-----------------------|
      | 00001011                 | 0 (padding)           |

    * Value Buffer:

      | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-63           |
      |-------------|-------------|-------------|-------------|-----------------------|
      | 1           | 2           | unspecified | 4           | unspecified (padding) |
```

File: arrow-ord/src/cmp.rs (hunk 3 above)
Comment (my-vegetable-has-exploded) [reply arrow-ord/src/cmp.rs:332]: get it.

**Comment by tustvold** [comment]
Correct, but the semantic of these kernels is any comparison against a null results in null output for that position

**Comment by my-vegetable-has-exploded** [comment]
> Can you please add tests for
> 
>     1. `distinct` / `not_distinct`
> 
>     2. A negative test that some operation like `lt` or `lt_eq` returns an error (not a panic) for struct arrays?
> 
>     3. A negative test that a struct array like `{a: int, b:int}` doesn't return `true` when compared to a struct array with a prefix like `{a:int}

Sure.




> Correct, but the semantic of these kernels is any comparison against a null results in null output for that position

I was wondering if it would be better to use ```Op::NotDistinct```  to check each field? More precisely, we need to go through the process in https://github.com/apache/arrow-rs/blob/9e060dc6c68e6aac30d0ae643d1d1854db44a378/arrow-ord/src/cmp.rs#L226-L284 after getting BooleanBuffer.

**Comment by my-vegetable-has-exploded** [comment]
> > I was wondering if it would be better to use Op::NotDistinct to check each field?
> 
> That would be a different kernel then. We definitely could/should support distinct/not_distinct for StructArray also, the difference with standard equality is how nulls are handled. Distinct follow the intuitive notions of equality, the equality kernels follow the SQL formulation of equality and the somewhat perverse null semantics it has 😅
> 
> https://learn.microsoft.com/en-us/sql/t-sql/queries/is-distinct-from-transact-sql?view=sql-server-ver16#remarks

I feel like I have caught your drift this time. Because the comparison between None and any value is Unknown, So {null, 2} is also not comparable. Thanks, I will change my code based on this suggestion.

File: arrow-ord/src/cmp.rs [hunk 7]
```diff
@@ -215,76 +299,114 @@ fn compare_op(op: Op, lhs: &dyn Datum, rhs: &dyn Datum) -> Result<BooleanArray,
...
+fn compare_op_struct_values(
+    op: Op,
+    l: &dyn Array,
+    l_s: bool,
+    r: &dyn Array,
+    r_s: bool,
+    len: usize,
+) -> Result<BooleanBuffer, ArrowError> {
+    // when one of field is equal, the result is false for not equal
+    // so we use neg to reverse the result of equal when handle not equal
```
Comment (tustvold) [inline arrow-ord/src/cmp.rs:384]: Why not just pass the operator into compare_op_values?

File: arrow-ord/src/cmp.rs [hunk 8]
```diff
@@ -215,76 +299,114 @@ fn compare_op(op: Op, lhs: &dyn Datum, rhs: &dyn Datum) -> Result<BooleanArray,
...
+
+    let l = l.as_struct();
+    let r = r.as_struct();
+
+    // compare each field of struct
+    let child_values = l
+        .columns()
+        .iter()
+        .zip(r.columns().iter())
+        .map(|(col_l, col_r)| compare_op_values(Op::Equal, col_l, l_s, col_r, r_s, len))
```
Comment (tustvold) [inline arrow-ord/src/cmp.rs:399]: I don't think this will correctly handle the null masks for a Distinct?

File: arrow-ord/src/cmp.rs [hunk 9]
```diff
@@ -702,4 +826,137 @@ mod tests {
...
+    #[test]
+    fn test_struct_compare() {
+        // test struct('a', 'b')、struct('a', 'b'), the null buffer is 0b0111
+        // left b[2] is different from right b[2]
+        let left_a = Arc::new(Int32Array::new(
+            vec![0, 1, 2, 3].into(),
+            Some(vec![true, false, true, true].into()),
+        ));
+        let right_a = Arc::new(Int32Array::new(
+            vec![0, 1, 2, 3].into(),
```
Comment (tustvold) [inline arrow-ord/src/cmp.rs:867]:
```suggestion
            vec![0, 72, 2, 3].into(),
```

This helps verify the null mask comparison is correct

File: arrow-ord/src/cmp.rs [hunk 10]
```diff
@@ -702,4 +826,137 @@ mod tests {
...
+        let field_a = Arc::new(Field::new("a", DataType::Int32, true));
+        let field_b = Arc::new(Field::new("b", DataType::Int32, true));
+        let left_struct = StructArray::from((
+            vec![
+                (field_a.clone(), left_a.clone() as ArrayRef),
+                (field_b.clone(), left_b.clone() as ArrayRef),
+            ],
+            Buffer::from([0b0111]),
+        ));
+        let right_struct = StructArray::from((
```
Comment (tustvold) [inline arrow-ord/src/cmp.rs:887]:
```suggestion
        // right [{a: 0, b: 0}, {a: NULL, b: 1}, {a: 2, b: 2}, {a: 3, b: 3} ]
        let right_struct = StructArray::from((
```

File: arrow-ord/src/cmp.rs [hunk 11]
```diff
@@ -702,4 +826,137 @@ mod tests {
...
+            vec![0, 1, 20, 3].into(),
+            Some(vec![true, true, true, true].into()),
+        ));
+        let right_b = Arc::new(Int32Array::new(
+            vec![0, 1, 2, 3].into(),
+            Some(vec![true, true, true, true].into()),
+        ));
+        let field_a = Arc::new(Field::new("a", DataType::Int32, true));
+        let field_b = Arc::new(Field::new("b", DataType::Int32, true));
+        let left_struct = StructArray::from((
```
Comment (tustvold) [inline arrow-ord/src/cmp.rs:880]:
```suggestion
        // [{a: 0, b: 0}, {a: NULL, b: 1}, {a: 2, b: 20}, {a: 3, b: 3}]
        let left_struct = StructArray::from((
```

**Review by tustvold — COMMENTED** [review]
Had a review, I like where this is headed but I don't think the null mask handling is quite right yet.

FWIW I'm not sure that separating out the null mask and values comparison makes sense, instead I would expect the logic to just recurse across the fields and union the null masks of the results (if any).

**Comment by my-vegetable-has-exploded** [comment]
> Had a review, I like where this is headed but I don't think the null mask handling is quite right yet.
> 
> FWIW I'm not sure that separating out the null mask and values comparison makes sense, instead I would expect the logic to just recurse across the fields and union the null masks of the results (if any), with a little bit of extra logic to handle any null mask in the struct array proper.

I wanted to do that at first. The main reason is that I found it's hard to reuse those codes

https://github.com/apache/arrow-rs/blob/3cd6da0a6b97056f3717d3491c84ec6d734aa262/arrow-ord/src/cmp.rs#L225-L285

If there is a better way to organize those codes, I'd like to have a try! Thanks a lot!

**Comment by tustvold** [comment]
It should be possible to just call compare_op recursively

**Comment by tustvold** [comment]
I'll have a play later today/tomorrow and see if I can't simplify this a bit

**Comment by my-vegetable-has-exploded** [comment]
> I'll have a play later today/tomorrow and see if I can't simplify this a bit

Thanks a lot，I’m sorry to add to your workload.

**Comment by Jefffrey** [comment]
Hey @tustvold & @my-vegetable-has-exploded , do we know the status of this PR now? It's been open for a bit and it seems there has been another PR for the same issue in the meantime, #5423, so wondering if efforts should be focused on a single PR? Otherwise can keep both open but mark this as draft as there hasn't been movement for a bit?

**Comment by tustvold** [comment]
Sorry this is partly on me, I'm somewhat struggling to keep up with all the various things going on. I think my preference is towards something along the lines of https://github.com/apache/arrow-rs/pull/5672 which would allow us to handle StructArray more comprehensively in the comparison kernels, instead of having non-trivial logic just for the case of equality.

I think let's mark this as a draft and I will try to find sometime next week to sort something out in this space

**Pull request closed without merging**